### ISSUE TITLE
Proxy logging request headers to error.log

### ISSUE THREAD

**Issue by initiateit** [litespeedtech/openlitespeed#166]
On the advice of a rep from the GoLiteSpeed slack channel I am logging this.

Whenever you setup an external app and configure a proxy, OLS seems to fill up the server error.log with request headers. If you have an app that constantly refreshes web sockets for data multiple times every few seconds then the log file amasses thousands of request headers every hour. It doesn't matter if you setup a rewrite rule and set the log level to 0, they just keep coming.

They look like this;

```
GET /api/3/all/views HTTP/1.1
host: admin.initiateit.com.au
accept: application/json, text/plain, */*
sec-fetch-dest: empty
user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.116 Safari/537.36
dnt: 1
sec-fetch-site: same-origin
sec-fetch-mode: cors
referer: https://admin.initiateit.com.au/
accept-encoding: gzip
accept-language: en-US,en;q=0.9
cookie: litespeed_admin_lang=english; __cfduid=d82860654aff8cfb6185034850993bae21582068187; LSUI37FE0C43B84483E0=5959e44169e5989300f53900dcb0d75a; LSID37FE0C43B8448$X-Forwarded-Host: admin.initiateit.com.au
X-Forwarded-Proto: https
X-Forwarded-For: 31.242.144.1
```

My proxy is  setup via an external app with the following rewrite rule;
```
Rewrite Rules:
RewriteRule /(.*)$ http://glances/$1 [P,E=Proxy-Host:admin.initiateit.com.au]
```
All logging is set to **ERROR** at both server config, vhost / proxy level.

The issue is not only are these not errors but they quickly fill up space on the disk but they consume disk writes for no purposeful reason. They also kill the formatting of the log file in the web admin panel and obscure log events as they dont fit the template. I suggest the following;

1. **Remove them** from being logged. I couldn't see where the logging in OLS dictates them being created so I couldn't say how to do it but a last resort of redirecting to /dev/null wouldn't bother me.
2. **Allow them** to be logged to a file of our choice, something like proxy.log - a selectable option in the web admin like is currently used for the logging.

Thanks for looking.

**Comment by h0bd0b1in** [comment]
We are observing the same thing with our setup, OLS 1.6.7 running on Ubuntu 18.04 LTS as a reverse proxy. As it stands, this is leading to us generating almost 250MB of error logs per day! What's worse, these entries appear in both the server error log and the virtual host error log (if one is set).

As a note, none of these headers show up in the admin error log, presumably because no proxying occurs here. Thus, I think this is related entirely to virtual hosts configured with a proxied External App. We have no non-proxied virtual hosts to test this theory on, though.

A potential side effect of this is that the OLS built-in log viewer has become unusable as it is unable to correctly parse the error log since these entries do not fit the log structure. This is speculation on my part based on the fact that we are unable to use the log viewer in the OLS admin interface.

Here is our (redacted) `httpd_config.conf`:

```
serverName                xxx
user                      nobody
group                     nogroup
priority                  0
cpuAffinity               1
enableLVE                 0
inMemBufSize              500M
swappingDir               /tmp/lshttpd/swap
autoFix503                1
enableh2c                 0
gracefulRestartTimeout    300
mime                      conf/mime.properties
disableInitLogRotation    1
showVersionNumber         2
enableIpGeo               1
useIpInProxyHeader        1
adminEmails               xxxxxxxxxx

errorlog $SERVER_ROOT/logs/error.log {
  logLevel                DEBUG
  debugLevel              0
  rollingSize             10M
  enableStderrLog         1
}

accesslog $SERVER_ROOT/logs/access.log {
  logHeaders              7
  rollingSize             10M
  keepDays                30
  compressArchive         1
}

indexFiles                index.html, index.php
autoIndex                 0

expires  {
  enableExpires           1
  expiresByType           image/*=A604800,text/css=A604800,application/x-javascript=A604800,application/javascript=A604800,font/*=A604800,application/x-font-ttf=A604800
}

autoLoadHtaccess          1

geoipdb /usr/share/GeoIP/GeoLite2-Country.mmdb {
  geoipDBName             COUNTRY_DB
}

geoipdb /usr/share/GeoIP/GeoLite2-City.mmdb {
  geoipDBName             CITY_DB
}

tuning  {
  maxConnections          10000
  maxSSLConnections       10000
  connTimeout             300
  maxKeepAliveReq         10000
  keepAliveTimeout        5
  sndBufSize              0
  rcvBufSize              0
  maxReqURLLen            16384
  maxReqHeaderSize        16380
  maxReqBodySize          5118M
  maxDynRespHeaderSize    32768
  maxDynRespSize          5118M
  maxCachedFileSize       4096
  totalInMemCacheSize     20M
  maxMMapFileSize         256K
  totalMMapCacheSize      40M
  useSendfile             1
  fileETag                28
  enableGzipCompress      1
  compressibleTypes       text/*, application/x-javascript, application/xml, application/javascript, image/svg+xml,application/rss+xml
  enableDynGzipCompress   1
  gzipCompressLevel       6
  gzipAutoUpdateStatic    1
  gzipStaticCompressLevel 6
  brStaticCompressLevel   6
  gzipMaxFileSize         10M
  gzipMinFileSize         300
  sslStrongDhKey          1
  sslSessionCache         1
}

fileAccessControl  {
  followSymbolLink        1
  checkSymbolLink         0
  requiredPermissionMask  000
  restrictedPermissionMask 000
}

perClientConnLimit  {
  staticReqPerSec         0
  dynReqPerSec            0
  outBandwidth            0
  inBandwidth             0
  softLimit               10000
  hardLimit               10000
  gracePeriod             15
  banPeriod               300
}

CGIRLimit  {
  maxCGIInstances         20
  minUID                  11
  minGID                  10
  priority                0
  CPUSoftLimit            10
  CPUHardLimit            50
  memSoftLimit            1460M
  memHardLimit            1470M
  procSoftLimit           400
  procHardLimit           450
}

accessDenyDir  {
  dir                     /
  dir                     /etc/*
  dir                     /dev/*
  dir                     conf/*
  dir                     admin/conf/*
}

accessControl  {
  allow                   ALL
}

extprocessor lsphp {
  type                    lsapi
  address                 uds://tmp/lshttpd/lsphp.sock
  maxConns                10
  env                     PHP_LSAPI_CHILDREN=10
  env                     LSAPI_AVOID_FORK=200M
  initTimeout             60
  retryTimeout            0
  persistConn             1
  respBuffer              0
  autoStart               1
  path                    fcgi-bin/lsphpnew
  backlog                 100
  instances               1
  priority                0
  memSoftLimit            2047M
  memHardLimit            2047M
  procSoftLimit           1400
  procHardLimit           1500
}

scripthandler  {
  add                     lsapi:lsphp php
}

railsDefaults  {
  maxConns                1
  env                     LSAPI_MAX_IDLE=60
  initTimeout             60
  retryTimeout            0
  pcKeepAliveTimeout      60
  respBuffer              0
  backlog                 50
  runOnStartUp            3
  extMaxIdleTime          300
  priority                3
  memSoftLimit            2047M
  memHardLimit            2047M
  procSoftLimit           500
  procHardLimit           600
}

wsgiDefaults  {
  maxConns                5
  env                     LSAPI_MAX_IDLE=60
  initTimeout             60
  retryTimeout            0
  pcKeepAliveTimeout      60
  respBuffer              0
  backlog                 50
  runOnStartUp            3
  extMaxIdleTime          300
  priority                3
  memSoftLimit            2047M
  memHardLimit            2047M
  procSoftLimit           500
  procHardLimit           600
}

nodeDefaults  {
  maxConns                5
  env                     LSAPI_MAX_IDLE=60
  initTimeout             60
  retryTimeout            0
  pcKeepAliveTimeout      60
  respBuffer              0
  backlog                 50
  runOnStartUp            3
  extMaxIdleTime          300
  priority                3
  memSoftLimit            2047M
  memHardLimit            2047M
  procSoftLimit           500
  procHardLimit           600
}

module cache {
  internal                1
#          1
checkPrivateCache   1
checkPublicCache    1
maxCacheObjSize     10000000
maxStaleAge         200
qsCache             1
reqCookieCache      1
respCookieCache     1
ignoreReqCacheCtrl  1
ignoreRespCacheCtrl 0
enableCache         0
expireInSeconds     3600
enablePrivateCache  0
privateExpireInSeconds 3600
  ls_enabled              1
}

virtualhost Example {
  vhRoot                  Example/
  configFile              conf/vhosts/Example/vhconf.conf
  allowSymbolLink         1
  enableScript            1
  restrained              1
  setUIDMode              0
}

virtualhost vhostA {
  vhRoot                  $SERVER_ROOT/vhostA
  configFile              $SERVER_ROOT/conf/vhosts/$VH_NAME/vhconf.conf
  allowSymbolLink         1
  enableScript            1
  restrained              1
}

virtualhost vhsotB {
  vhRoot                  $SERVER_ROOT/vhostB
  configFile              $SERVER_ROOT/conf/vhosts/$VH_NAME/vhconf.conf
  allowSymbolLink         1
  enableScript            1
  restrained              1
}

virtualhost vhostC {
  vhRoot                  $SERVER_ROOT/vhostC
  configFile              $SERVER_ROOT/conf/vhosts/$VH_NAME/vhconf.conf
  allowSymbolLink         1
  enableScript            1
  restrained              1
}

virtualhost vhostD {
  vhRoot                  $SERVER_ROOT/vhostD
  configFile              $SERVER_ROOT/conf/vhosts/$VH_NAME/vhconf.conf
  allowSymbolLink         1
  enableScript            1
  restrained              1
}

virtualhost vhostE {
  vhRoot                  $SERVER_ROOT/vhostE
  configFile              $SERVER_ROOT/conf/vhosts/$VH_NAME/vhconf.conf
  allowSymbolLink         1
  enableScript            1
  restrained              1
}

virtualhost vhostF {
  vhRoot                  $SERVER_ROOT/vhostF
  configFile              $SERVER_ROOT/conf/vhosts/vhostF/vhconf.conf
  note                    This is the Virtual Host configuration for maintenance mode
  allowSymbolLink         1
  enableScript            0
  restrained              1
}

virtualhost vhostG {
  vhRoot                  vhostG/
  configFile              $SERVER_ROOT/conf/vhosts/$VH_NAME/vhconf.conf
  allowSymbolLink         1
  enableScript            0
  restrained              1
}

virtualhost vhostH {
  vhRoot                  $SERVER_ROOT/vhostH
  configFile              $SERVER_ROOT/conf/vhosts/$VH_NAME/vhconf.conf
  allowSymbolLink         1
  enableScript            1
  restrained              1
}

listener Default {
  address                 *:80
  secure                  0
  map                     SSLRedirect *
}

listener Defaultssl {
  address                 *:443
  secure                  1
  keyFile                 /usr/local/lsws/conf/example.key
  certFile                /usr/local/lsws/conf/example.crt
  map                     xxxxxxxx
  map                     xxxxxxxx
  map                     xxxxxxxx
  map                     xxxxxxxx
  map                     xxxxxxxx
  map                     xxxxxxxx
}

vhTemplate centralConfigLog {
  templateFile            conf/templates/ccl.conf
  listeners               Default
}

vhTemplate EasyRailsWithSuEXEC {
  templateFile            conf/templates/rails.conf
  listeners               Default
}

suspendedVhosts           vhostA,vhostF
```

And an example vhost file:

```
docRoot                   $VH_ROOT/html
vhDomain                  xxxxxx
enableGzip                1
enableIpGeo               1
cgroups                   0

errorlog $VH_ROOT/logs/error.log {
  useServer               0
  logLevel                ERROR
  rollingSize             8M
}

accesslog $VH_ROOT/logs/access.log {
  useServer               0
  logHeaders              7
  rollingSize             8M
  keepDays                14
  compressArchive         1
}

index  {
  useServer               1
  autoIndex               0
}

extprocessor ExtAppA {
  type                    proxy
  address                 <removed>
  maxConns                100
  initTimeout             120
  retryTimeout            3
  respBuffer              0
}

extprocessor ExtAppB {
  type                    proxy
  address                 <removed>
  maxConns                2000
  initTimeout             120
  retryTimeout            3
  respBuffer              0
}

context / {
  type                    proxy
  handler                 ExtAppA
  extraHeaders            Header set Strict-Transport-Security "max-age=63072000; includeSubDomains; preload"
  addDefaultCharset       off
  enableIpGeo             1
}

rewrite  {
  enable                  1
  rules                   <<<END_rules
RewriteEngine On
RewriteRule <removed>
RewriteRule <removed>
RewriteRule <removed> [END]
RewriteRule <regex removed> "http://ExtAppB/$1" [P]
  END_rules
}

vhssl  {
  keyFile                 $VH_ROOT/certs/privkey.pem
  certFile                $VH_ROOT/certs/fullchain.pem
  certChain               1
  sslProtocol             24
  DHParam                 $SERVER_ROOT/dhparam.pem
  enableSpdy              12
}
```

Hopefully that's enough to give the devs something to work with to resolve the issue.

**Comment by initiateit** [comment]
Can we get an update on this please? This is not currently assigned and the issue is a pain in the proverbial to have to deal with.

**Comment by takerukoushirou** [comment]
I ran into the same issue, the error logs are practically useless and become excessively large if OLS is used as a reverse proxy.

Looking at the source code, I guess that the headers were probably supposed to be some debug output, but are actually always logged, see [line 337 of proxyconn.cpp](https://github.com/litespeedtech/openlitespeed/blob/master/src/extensions/proxy/proxyconn.cpp#L337). Unlike other debug out that is wrapped within the `LS_DBG_L` macro, the headers are written verbatim to the log (which also explains why the line endings of the logged headers are actually CR+LF instead of standard UNIX LF-only).
Commenting the lines out or wrapping them in a debug log macro might fix the issue (couldn't test it myself due to time constraints).

According to the repository history, these changes were introduced with the [1.5.1 check-in](https://github.com/litespeedtech/openlitespeed/commit/7cfcda1f3cb2bafa24fc4aa5bc448071530d1536#diff-2e1b9a95bdd873f1ece88885ee16cb1ef9521902f70dc21ebb7f55a7a0927600).

**Comment by litespeedtech** [comment]
Thanks for looking into this. 
Yes, you are correct. it should be a conditional log, follows debug log level. Will have it fixed.

**Comment by litespeedtech** [comment]
Pre-release packages for 1.7.9, 1.6.20 are available now. Can use follow command to apply those fixes
`/usr/local/lsws/admin/misc/lsup.sh -b -e 1.7.9`
or
`/usr/local/lsws/admin/misc/lsup.sh -b -e 1.6.20`

**Comment by litespeedtech** [comment]
It has been fixed in 1.6.20 and 1.7.9 release.
